policy-engine/src/graph.rs

```diff
@@ -35,93 +34,77 @@ pub(crate) fn register_metrics(registry: &Registry) -> Fallible<()> {
-        .mandatory_params;
-    if let Err(e) = commons::ensure_query_params(mandatory_params, req.query_string()) {
-        return Box::new(future::err(e));
-    }
+    let mandatory_params = &app_data.mandatory_params;
+    commons::ensure_query_params(mandatory_params, req.query_string())?;
 
-    let plugin_params = match Query::<HashMap<String, String>>::from_query(req.query_string()) {
-        Ok(query) => query.into_inner(),
-        Err(e) => {
-            return Box::new(futures::future::err(commons::GraphError::InvalidParams(
-                e.to_string(),
-            )))
-        }
-    };
+    let plugin_params = Query::<HashMap<String, String>>::from_query(req.query_string())
+        .map(|query| query.into_inner())
+        .map_err(|e| commons::GraphError::InvalidParams(e.to_string()))?;
 
-    let plugins = req
-        .app_data::<AppState>()
-        .expect(commons::MISSING_APPSTATE_PANIC_MSG)
-        .plugins;
+    let plugins = &app_data.plugins;
 
     let timer = V1_GRAPH_SERVE_HIST.start_timer();
-    let serve = futures::future::ok(())
-        .and_then(move |_| {
-            cincinnati::plugins::process(
-                plugins.iter(),
-                cincinnati::plugins::PluginIO::InternalIO(cincinnati::plugins::InternalIO {
-                    graph: Default::default(),
-                    parameters: plugin_params,
-                }),
-            )
-            .map_err(|e| match e.downcast::<GraphError>() {
-                Ok(graph_error) => graph_error,
-                Err(other_error) => GraphError::FailedPluginExecution(other_error.to_string()),
-            })
-        })
-        .and_then(|internal_io| {
-            serde_json::to_string(&internal_io.graph)
-                .map_err(|e| GraphError::FailedJsonOut(e.to_string()))
-        })
-        .map(|graph_json| {
-            HttpResponse::Ok()
-                .content_type(CONTENT_TYPE)
-                .body(graph_json)
-        })
-        .then(move |r| {
-            timer.observe_duration();
-
-            if let Err(e) = &r {
-                error!(
-                    "Error serving request '{}' from '{}': {:?}",
-                    format!("{:?}", &req).replace("\n", " ").replace("\t", " "),
-                    &req.peer_addr()
-                        .map(|addr| addr.to_string())
-                        .unwrap_or("<not available>".into()),
-                    e
-                );
-            }
 
-            r
-        });
-    Box::new(serve)
+    let response = cincinnati::plugins::process(
+        plugins.iter(),
+        cincinnati::plugins::PluginIO::InternalIO(cincinnati::plugins::InternalIO {
+            graph: Default::default(),
+            parameters: plugin_params,
+        }),
+    )
+    .await
+    .map_err(|e| match e.downcast::<GraphError>() {
+        Ok(graph_error) => graph_error,
+        Err(other_error) => GraphError::FailedPluginExecution(other_error.to_string()),
+    })
```

Comment:
I'd suggest breaking the chain here and moving the rest of the logic below out of combinators.

Reply:
That will increase the effort of logging the error from anywhere in the chain, as we do it now at the end of the chain. Unless you have a suggestion how to retain that mechanism somehow?

Reply:
I see. Then fell free to keep it. An alternative could be to put the processing logic in its own function, and the error-logging logic as a wrapper one level above it.

Reply:
I implemented your suggestion in a fixup commit. PTAL and let me know if you like that better in terms of readability. If so, I'm fine squashing it.

Reply:
That looks good to me, but I know it's subjective taste. Leaving the final call to you.

Reply:
I squashed the commit into the rest.